eos-tech-support/eos-download-image

```diff
@@ -1,125 +1,114 @@
-  exit 1
-fi
-
-# Download & verify
-download_and_verify "image" $BASE_URL/$OS_IMAGE_URL $BASE_URL/$OS_SIGNATURE_URL >&2
-
-echo $(basename $OS_IMAGE_URL)
+#!/usr/bin/env python3
+# vim: fileencoding=utf-8 sts=4 sw=4 et
+import argparse
+import codecs
+import distutils.version
+import gzip
+import json
+import os
+import pipes
+import subprocess
+import sys
+import urllib.request
+
+INTERNAL_BASE_URL = "http://images.endlessm-sf.com"
+PUBLIC_BASE_URL = "https://d1anzknqnc1kmb.cloudfront.net"
+KEYRING_FILENAME = "eos-image-keyring.gpg"
+KEYRING_URL = PUBLIC_BASE_URL + "/" + KEYRING_FILENAME
+KEYRING_SYSTEM_PATH = "/usr/share/keyrings/eos-image-keyring.gpg"
+
+
+def log(*args):
+    sys.stderr.write('\n')
+    print(*args, file=sys.stderr)
+    sys.stderr.flush()
+
+
+def check_call(*args):
+    log('$ ' + ' '.join(map(pipes.quote, args)))
+    subprocess.check_call(args, stdout=sys.stderr)
+
+
+def ensure_keyring():
+    if os.path.isfile(KEYRING_SYSTEM_PATH):
+        return KEYRING_SYSTEM_PATH
+
+    keyring = os.path.abspath(KEYRING_FILENAME)
+    if not os.path.isfile(keyring):
+        log("Fetching Endless keyring")
+        check_call("wget", "-O", keyring, KEYRING_URL)
+
+    return keyring
+
+
+def download_and_verify(image_url, signature_url):
+    log("Downloading image from", image_url)
+    check_call("wget", "--continue", image_url)
+
+    log("Downloading signature from", signature_url)
+    check_call("wget", "--continue", signature_url)
+
+    keyring = ensure_keyring()
+
+    log("Verifying")
+    check_call("gpgv", "--keyring", keyring,
+               os.path.basename(signature_url),
+               os.path.basename(image_url))
+
+    log("Download okay.")
+    return os.path.basename(image_url)
+
+
+def main():
+    p = argparse.ArgumentParser(
+        description='Fetches and verifies an image (installer or regular OS)')
+    p.add_argument('-v', '--version',
+                   help='Image version (default: newest)')
+    p.add_argument('-p', '--personality',
+                   help='Image personality (default: base)',
+                   default='base')
+    p.add_argument('-r', '--product',
+                   help='Product (eg: eosinstaller; default: eos)',
+                   default='eos')
+    p.add_argument('-i', '--internal',
+                   action='store_true',
+                   help='Fetch images from the Endless internal network')
+
+    args = p.parse_args()
+    base = INTERNAL_BASE_URL if args.internal else PUBLIC_BASE_URL
+
+    manifest_url = '{base}/releases-{product}.json.gz'.format(
+        base=base, product=args.product)
+    req = urllib.request.Request(manifest_url)
+    with urllib.request.urlopen(req) as conn:
+        with gzip.GzipFile(mode='r', fileobj=conn) as manifest_fp:
+            manifest = json.load(codecs.getreader('utf8')(manifest_fp))
+
+    images = list(manifest['images'].values())
+    images.sort(key=lambda i: distutils.version.StrictVersion(i['version']))
```

Comment:
I wish we could use apt versioning here since that's what we use everywhere else, but it looks like `python3-apt` is not part of the core package list.